### PR TITLE
Add function body for SoftmaxCrossEntropyLossGrad

### PR DESCRIPTION
**Pull request by gramalingam**
**Description**: 

Add function body for SoftmaxCrossEntropyLossGrad

### CODE REVIEW

File: onnxruntime/test/contrib_ops/function_test_util.cc [hunk 1]
```diff
@@ -147,7 +155,7 @@ std::unique_ptr<Model> FunctionTestCase::CreateModel(bool inline_call) {
 
   if (inline_call) {
     EXPECT_STATUS_OK(graph.InlineFunction(call_node));
-#if 0
+#if 1
```
Comment (aaronsm) [inline onnxruntime/test/contrib_ops/function_test_util.cc]: #if 0?

File: onnxruntime/test/contrib_ops/function_test_util.h [hunk 2]
```diff
@@ -23,6 +23,19 @@ inline std::vector<T> random(std::vector<int64_t> shape) {
...
+
+template <>
+inline std::vector<int64_t> random<int64_t>(std::vector<int64_t> shape) {
+  int64_t size = 1;
+  for (auto dim : shape)
+    size *= dim;
+
+  std::vector<int64_t> data(size);
+  for (int64_t i = 0; i < size; i++)
+    data[i] = static_cast<int64_t>(rand());
```
Comment (edgchen1) [inline onnxruntime/test/contrib_ops/function_test_util.h:35]: using rand() will bypass the test random seed, consider using RandomValueGenerator instead

File: onnxruntime/test/contrib_ops/function_test_util.h [hunk 3]
```diff
@@ -88,9 +101,23 @@ struct FunctionTestCase {
     }
   }
 
+  template <typename T>
+  void AddBoundedInput(std::string input_name, std::vector<int64_t> shape, T bound) {
```
Comment (edgchen1) [inline onnxruntime/test/contrib_ops/function_test_util.h:105]: const string&, const vector&

File: orttraining/orttraining/core/graph/training_op_defs.cc [hunk 4]
```diff
@@ -1447,6 +1447,132 @@ Example 4:
...
+            FunctionBuilder builder(functionProto);
+
+            // Inputs:
+            // dY : scalar (if reduced) or [B, d1, d2, ...]
+            // log_prob : [B, C, d1, d2, ...]
+            // weight : [C]
+            // label : [B, d1, d2, ...]
+
+            // We decompose the forward propagation into two steps, for doing the backward prop.
+            // Step 1: loss = Neg(Logsoftmax(prediction-for-true-label))
```
Comment (codemzs) [inline orttraining/orttraining/core/graph/training_op_defs.cc:1469]: >Neg(Logsoftmax(prediction-for-true-label)) [](http://example.com/codeflow?start=30&length=42)

Negative Log-likelihood(NLL)?

**Comment by codemzs** [comment]
>   ONNX_CONTRIB_OPERATOR_SCHEMA(SoftmaxCrossEntropyLossGrad)

The node we see in the graph is SoftmaxCrossEntropyLossInternalGrad, do we need to make this change there as well? Same for SoftmaxCrossEntropyLossInternal

---
Refers to: orttraining/orttraining/core/graph/training_op_defs.cc:1419 in 0cfa560. [](commit_id = 0cfa560cb3151438346c7295294f6aaede20d38f, deletion_comment = False)

File: orttraining/orttraining/core/graph/training_op_defs.cc [hunk 5]
```diff
@@ -142,13 +142,189 @@ TensorProto ToDimensionOneFloatTensor(float value) {
...
+                log_prob_BCD = Reshape (log_prob, reshape_arg)
+                prob_BCD = Exp (log_prob_BCD)
+
+                label_BD = Flatten (label) # convert from [B, d1, d2, ...] to [B, D = d1 * d2 * ...]
+
+                zero_one = Constant < value = int32[2] {0, 1}>()
+                zero_one_typed = CastLike (zero_one, prob_BCD)
+                C1d = Shape <start = 1, end = 2> (prob_BCD)
+                C = Squeeze(C1d)
+                one_hot_label_BCD = OneHot <axis=1> (label_BD, C, zero_one_typed)
```
Comment (codemzs) [inline orttraining/orttraining/core/graph/training_op_defs.cc:315]: >OneHot [](http://example.com/codeflow?start=36&length=6)

Is there a less scary alternative to this that you can think of?

**Review by souptc — APPROVED** [review]

File: orttraining/orttraining/core/graph/training_op_defs.cc [hunk 6]
```diff
@@ -142,13 +142,189 @@ TensorProto ToDimensionOneFloatTensor(float value) {
...
+  // by the reshaping for the general case.
+  builder.Add(R"(
+                d_loss_B1Dopt = Unsqueeze (d_loss, axes1)
+                reshape_arg = Constant < value = int64[3] {0, 0, -1} > ()
+                d_loss_B1D = Reshape (d_loss_B1Dopt, reshape_arg)
+                orig_shape = Shape (log_prob)
+                log_prob_BCD = Reshape (log_prob, reshape_arg)
+                prob_BCD = Exp (log_prob_BCD)
+
+                label_BD = Flatten (label) # convert from [B, d1, d2, ...] to [B, D = d1 * d2 * ...]
```
[the remark this answers is not in the document]
Comment (codemzs) [reply orttraining/orttraining/core/graph/training_op_defs.cc:309]: I'm curious what ops will use to get the product D = d1*d2...?

[the remark this answers is not in the document]
Comment (gramalingam) [reply orttraining/orttraining/core/graph/training_op_defs.cc:309]: Does it really matter? `Flatten` is just a simpler shorthand for a two-line `Reshape`, and easier to analyze, etc. The CI takes so much time, and some of those fail randomly. If there is a good reason, I can change it, but if not, let's merge this.

File: orttraining/orttraining/core/graph/training_op_defs.cc (hunk 5 above)
Comment (gramalingam) [reply orttraining/orttraining/core/graph/training_op_defs.cc:315]: Not really. There is an alternative encoding which is a bit longer, and which will probably make efficient compilation harder, so not sure if would be helpful. What metric would you use to choose one as being better (or scarier) than another?

File: orttraining/orttraining/core/graph/training_op_defs.cc (hunk 6 above)
[the remark this answers is not in the document]
Comment (gramalingam) [reply orttraining/orttraining/core/graph/training_op_defs.cc]: Updated, to replace use of Flatten and OneHot.

**Review by souptc — APPROVED** [review]
